expirationd.lua

```diff
@@ -359,21 +363,119 @@ local function expirationd_run_task(name, space_id, is_tuple_expired, options)
     end
     task.process_expired_tuple = options.process_expired_tuple or default_tuple_drop
 
+    -- validate index
+    local expire_index = box.space[space_id].index[0]
+    if options.index then
+        local index_exists = false
+        for _, index in pairs(box.space[space_id].index) do
+            if index.name == options.index then
+                expire_index = index
+                index_exists = true
+            end
+        end
+        if not index_exists then
+            error("Index with name " .. options.index .. " not exists")
+        end
+        if expire_index.type ~= 'TREE' and expire_index.type ~= 'HASH' then
+            error("Not supported index type, expected TREE or HASH")
+        end
+    end
+    task.expire_index = expire_index
+
+    -- check iterator
+    if options.iterator ~= nil then
+        if type(options.iterator) ~= 'string' then
+            error("Invalid type of iterator, expected string")
+        end
+        if expire_index.type == 'TREE' then
+            if  options.iterator ~= 'ALL' and
+                    options.iterator ~= 'EQ'  and
+                    options.iterator ~= 'REQ' and
+                    options.iterator ~= 'GT'  and
+                    options.iterator ~= 'GE'  and
+                    options.iterator ~= 'LT'  and
+                    options.iterator ~= 'LE'  then
+                error("Invalid iterator for TREE index, expected {ALL, EQ, REQ, GT, GE, LT, LE}")
+            end
+        elseif expire_index.type == 'HASH' then
+            if  options.iterator ~= 'ALL' and
+                    options.iterator ~= 'EQ'  and
+                    options.iterator ~= 'GT'  then
+                error("Invalid iterator for HASH index, expected {ALL, EQ, GT}")
+            end
+        end
+        task.iterator = options.iterator
+    else
+        if expire_index.type == 'HASH' then
```

Comment:
There is a check for valid iterators for the hash index, since, according to the documentation, we can use three iterators ALL, EQ, GT

Reply:
You can validate iterator just call `index:pairs(key, iterator)`. Of course your comparison is incorrect since it's possible to use numeric value from `box.index.` e.g. `box.index.EQ`.

Why `index:pairs(key, iterator)` will work. Because this command only creates iterator but doesn't touch space anyhow. I hope you remember (or could read in Alexey's email about iterator stability) that iterators are lazy.
And if you can create iterator then it's correct.

Reply:
Made the check by calling index: pairs (key, iterator), as suggested by @olegrok

Reply:
NB: There is `box.internal.check_iterator_type(opts, key_is_nil)`. Example: https://github.com/tarantool/graphql.0/pull/260/files#diff-29d7b710a75ba3836cc3266be5b3ba6db6e20757047f54414b5c91e7ccb2e607R29-R37